[4.0] Child templates

dgrammatiko opened the pull request:
Pull Request is the foundation for #30149

### Summary of Changes
Introduces Inheritable templates solving numerous problem with the current state of templates. Read more on the linked issue above

This PR is complete in the sense that contains all the changes (even the upgrade path from J3) to introduce the new mode but without introducing the GUI and neither is moving any of the templates to the new mode. These can be done in following PR's


### Testing Instructions

Do not try to test this with patch tester, it won't work.
Install the PR's installable, you can d/l it here (at the bottom of this page)
If you want to use git command lines or the desktop app make sure that you'll do a fresh install or apply the updates (sql files)

Make sure you have the browser's console open so you can observe any missing assets.

If this went smooth, then it's a good sign as the installation is a seperate application with it's own template and although this PR is not touching the installation template everything works as before. B/C check 1

### Front End step 1

Go to the front end and vist the following urls:

- /index.php/blog
- /index.php/blog/3-welcome-to-your-blog
- /index.php/author-login (login here)
- /index.php?option=com_config&view=modules&id=109&Itemid=104 (or just try to edit the top left module)

Everything should be fine, no console logs or missinng icons, etc. B/C check 2

### Front End step 2

Download this [cassiopeia_overrides_v2.zip](https://github.com/joomla/joomla-cms/files/4977969/cassiopeia_overrides_v2.zip) and extract the contents and copy them in the folder `templates/cassiopeia/html`. Revisit the pages of the previous step. You should see in the pages some plain text (that will also break a bit the design) `Legacy Component view override`. You have to look for `Component`, `Module`, `Plugin` and `JLayout`. If so then the cascading of the templating is working correctly, in the previous step there was no overrides in this one the overrides were treated as top priority. Also cassiopeia has some overrides for the static assets so that was already tested (if you want you might open a normal 4.0 version in another winndow and check the number of assets per page between the 2). B/C check 3

### Front End step 3

Install the first new mode template (basically cassiopeia with another name: siteparent) from here [siteparent_v2.zip](https://github.com/joomla/joomla-cms/files/4977972/siteparent_v2.zip)


Go to the admin and make the siteparent template default and also assign all the menus to this template!!! Go to the front end and repeat the Front End step 1 and confirm that everything is as before. check 4

### Front End step 4

Go with your filemanager to `templates/siteparent` and copy the contents os the folder `xhtml` to the folder `html`. Redo the ritual of Front End step 1 and confirm that all the parts `Component`, `Module`, `Plugin` and `JLayout` are there. check 5

### Front End step 5

Although the child templates should be created inside joomla's env for testing purposes we will do it using the installer, so install this: [sitechild_v2.zip](https://github.com/joomla/joomla-cms/files/4977974/sitechild_v2.zip)



Go to the admin and make the sitechild template default and also assign all the menus to this template!!! Go to the front end and repeat the Front End step 1 and confirm that everything is as before. check 6

### Front End step 6

Go with your filemanager to `templates/sitechild` and rename the folder `xhtml` to `html`. Redo the ritual of Front End step 1 and confirm that all the parts `Component`, `Module`, `Plugin` and `JLayout` are there. check 7

### Back End step 1

Go to the addinistrator end and vist the following urls:

- /administrator/index.php
- /administrator/index.php?option=com_content&view=articles
- /administrator/index.php?option=com_installer&view=install

Everything should be fine, no console logs or missinng icons, etc. B/C check 8

### Back End step 2

Download this [atum_overrides_v2.zip](https://github.com/joomla/joomla-cms/files/4977975/atum_overrides_v2.zip) and extract the contents and copy them in the folder `administrator/templates/atum/html`. Revisit the pages of the previous step. You should see in the pages some plain text (that will also break a bit the design) `Legacy Component view override`. You have to look for `Component`, `Module`, `Plugin` and `JLayout`. If so then the cascading of the templating is working correctly, in the previous step there was no overrides in this one the overrides were treated as top priority. Also cassiopeia has some overrides for the static assets so that was already tested (if you want you might open a normal 4.0 version in another window and check the number of assets per page between the 2). B/C check 9

### Back End step 3

Install the new mode template (basically atum with another name: adminparent) from here 
[adminparent_v2.zip](https://github.com/joomla/joomla-cms/files/4977976/adminparent_v2.zip)


Go to the administrator template styles and make the adminparent template the default one. Repeat the Back End step 1 and confirm that everything is as before. check 10

### Back End step 4

Go with your filemanager to `administrator/templates/adminparent` and copy the contents os the folder `xhtml` to the folder `html`. Redo the dance of the Back End step 1 and confirm that all the parts `Component`, `Module`, `Plugin` and `JLayout` are there. check 11

### Back End step 5

Althouth the child templates should be created inside joomla's env for testing purposes we will do it using the installer, so install this: 
[adminchild_v2.zip](https://github.com/joomla/joomla-cms/files/4977978/adminchild_v2.zip)



Go to the admin and make the adminchild template the default. Repeat the Back End step 1 and confirm that everything is as before. check 12

### Back End step 6

Go with your filemanager to `administrator/templates/adminchild` and rename the folder `xhtml` to `html`. Redo the ritual of the Back End step 1 and confirm that all the parts `Component`, `Module`, `Plugin` and `JLayout` are there. check 13

### Final checks

Check your db and the table template_styles, you should have

- autum and cassiopeia with parent `0` and inherits ``
- from the new ones the parent ones with parent `1` and inherits ``
- and the child ones ones with parent `0` and inherits `siteparent` and `siteadmin` respectfully

Finally someone needs to confirm that the update sql files are all that is needed here

### And the last one

You have probably saw some missing strings when you were trying to edit any of the template styles. This is expected as I forgot to include any languages in the parent template, but this is quite easy to check. You are already in the adminchild template so all you have to do is create a folder `language/en-GB` in the **adminparent** template and then copy the tmpl_atum\*.ini to this folder and rename them to tmpl_adminparent.ini and tmpl_adminparent.sys.ini. Check again the strings are back. Then make the adminparent the default template and check for any changes (readable text should remain readable). You can repeat this for the front end templates (although the function is common for both).


### Documentation Changes Required
For the devs the new parts are 2:
- manifest **requires** a tag `<inheritable>1</inheritable >` to enable a template in the new mode
- **all** static assets **and also any static asset override** should be done in the `media/( administrator || site )/templateName`. Folders like css/js/images inside the above path will act exactly the same as they did in the old `templates/name/( css || js || images )`

dgrammatiko committed:
Sweet child o’mine

dgrammatiko committed:
Update base.sql

dgrammatiko committed:
CS

Fedik commented on libraries/src/Document/HtmlDocument.php:
```diff
@@ -804,15 +804,25 @@ protected function _fetchTemplate($params = array())
-		$template = $filter->clean($params['template'], 'cmd');
+		$templateName = $filter->clean($params['template'], 'cmd');
 		$file = $filter->clean($params['file'], 'cmd');
+		$activeTemplate = CmsFactory::getApplication()->getTemplateByName(true, $templateName);
```
I finally understood why you made this `getTemplateByName()` :smile: 
Well, this is hacking around, you actually do not need it.

Look for runtime how the template name gets here, it goes like:
```
SiteApplication::dispatch() 
SiteApplication::render() 
  => CMSApplication::render() 
    => HtmlDocument::parse($params)
       => HtmlDocument::_fetchTemplate($params)
```

Instead of all `getTemplateByName()`, you can add one more option to `$params`, and use it in `_fetchTemplate()`.

Look `SiteApplication::dispatch` where template first loaded:
https://github.com/joomla/joomla-cms/blob/bf5e54ca1e2cce162e116d55a1f0bc82e6381bda/libraries/src/Application/SiteApplication.php#L164-L170

Here the place where you have to set "style folder" (you called it "inherit"):
```php
$this->set('themeInherit', $template->Inherit);
```

Then look `CMSApplication::render` where Application sets Doc params:
https://github.com/joomla/joomla-cms/blob/d7451e644001e6fbdb85abf0c87c9c34bae2602c/libraries/src/Application/CMSApplication.php#L939-L946

Here you take $app `themeInherit` property (that was set in previous step) and set as Doc params:

```php
$this->docOptions['templateInherit']  = $this->get('themeInherit');
```

And finally, inside `HtmlDocument::_fetchTemplate()`, you can use new param `$params['templateInherit']`;

That all you need, no need this `getTemplateByName()`

dgrammatiko replied:
That's nice but there is no more case: `?template=blah` and in this one I have to fetch from db any params for the requested template (as it might be child and thus we need the inherits part form the db otherwise it will crash hard). Anyways I will review again the code so it might be unnecessary all this extra code

Fedik replied:
> but there is no more case: ?template=blah 

you have mixed the `template` name with the `template style` here, 
if you need `template style` you do `?templateStyle=<id>` to enable your style

This already exist in `SiteApplication::getTemplate()`
https://github.com/joomla/joomla-cms/blob/bf5e54ca1e2cce162e116d55a1f0bc82e6381bda/libraries/src/Application/SiteApplication.php#L426-L428

dgrammatiko replied:
> you have mixed the template name with the template style here

There were 2 options here:
- either add the parent and inherits columns to the extensions table 
- or add them to the template_styles.

 I choose the second thus the mixed logic in the getTemplateByName. If people don't object adding the info in the extensions table then all this weird part can be dropped.

Fedik replied:
yeah you did it as template style and this is good, but you do not need getTemplateByName() and do not need to add anything to extensions table. It like this:

to load **template** by request you do `?template=blah` 
to load **template style** (so your override) you do `?templateStyle=<style id>` (that what you need for test)

dgrammatiko replied:
> to load template by request you do ?template=blah

The `?template=blah` will be broken for a child template (eg just an xml file) **without** the db data from the `inherits` column which is stored in the styles table. I'm not an idiot I understand what you you're saying, but the (parent or child) templates **need** the information for both parent and inherits otherwise they cannot switch the logic in many places and so the app will fail spectacularly

dgrammatiko replied:
@Fedik hold my beer, I'll revert all these crap and do it in a clean way

Fedik replied:
> but the (parent or child) templates need the information for both parent and inherits otherwise they cannot switch the logic in many places

that why I wrote that big comment :wink: 

> I'll revert all these crap and do it in a clean way

I can make a pull request if need, just ping me when you revert that

dgrammatiko replied:
@Fedik can you review it now?

Fedik replied:
looks much better now, thanks!

Fedik commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -475,6 +497,9 @@ public function getTemplate($params = false)
 			foreach ($templates as &$template)
 			{
+				$template->inherits = !empty($template->inherits) ? $template->inherits : false;
+				$template->parent = $template->parent === 1;
```
please keep it as integer, if it is integer in database, 
because this is not safe, such change  may lead to much problem in future

dgrammatiko replied:
ok

kawshar commented:
My recommendation, please keep the database out of it. Just a normal override option would be great. I don't know why Joomla has deep integration with the database for all possible features except the most important Media Manager.

Thanks

dgrammatiko committed:
Remove uneeded code

Thank @Fedik for this!!

dgrammatiko committed:
typo

dgrammatiko commented:
> My recommendation, please keep the database out of it

One way or another the db is involved here as it holds the style data

> I don't know why Joomla has deep integration with the database

Very short answer: db is the only secure storage it got. Just don't ask me why the template style details (or any such data) needs any elevated security 😂

Fedik commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -751,6 +795,9 @@ protected function render()
 				}
 
+				// Pass the parent template to the state
+				$this->set('themeInherit', $template->inherits);
```
a typo, I think you meant `templateInherits`?  :wink:

dgrammatiko replied:
yup, fixed

kawshar replied:
> Very short answer: db is the only secure storage it got. Just don't ask me why the template style details (or any such data) needs any elevated security 😂

I don't know but Joomla is getting more difficult to use. For example, I have seen a new feature called "User Notes" which has categories integration. Could you please explain on what earth user notes need categories but such powerful cms media manager does not have any kind of database integration or ACL?

Btw, good luck man!

brianteeman replied:
User notes was introduced on 5 Dec 2011

dgrammatiko replied:
TIL `User Notes` need categories, wow the definitive new feature that nobody will ever be informed about or come close to using it!!!

>  cms media manager does not have any kind of database integration or ACL

I was involved with the first stage of Media Manager, which was basically a drop-in replacement for the current one but with a better API (eg support for editing, external storage, etc). AFAIK there was a stage 2 planned for db integration and some other missing parts but I was not part of it or there was never formed a user working group. So I know that this was the plan but I don't know why it didn't happen so far...

brianteeman replied:
Not new. Been there since 2011!!

dgrammatiko replied:
I stand corrected

kawshar replied:
> Not new. Been there since 2011!!

Good to know & never noticed :D

kawshar commented:
> Very short answer: db is the only secure storage it got. Just don't ask me why the template style details (or any such data) needs any elevated security 😂
I don't know. Everything is getting difficult day by day. I have seen a new feature in Joomla called "User Notes" and that has categories. On what earth you need a user note categories and a huge CMS totally has no database management of the media.

Good luck man!

Fedik commented:
Couple 2 note :)

**First note.**
I am not strong in English, but for me the property name `inherits` is a bit of controversial (Usually it means the children inherit parent.)

In current usage I would call it `style_template`, so the template style object will have 2 property:
Existing `template` the name of folder for parent template
New `style_template` the name of folder for override (child template).


**Second note.**
I am not very understood why need `parent` property.
I think if the template has empty `inherits` it is parent automatically, or maybe I missed something?

dgrammatiko commented:
> I am not very understood why need parent property.
Check these lines in the HTMLHelper:
```php
				$templaPath = JPATH_THEMES;

 				if ($template->parent || !empty($template->inherits))
 				{
 					$client     = $app->isClient('administrator') === true ? 'administrator' : 'site';
 					$templaPath = JPATH_ROOT . "/media/templates/$client";
 				}
```

Basically it's a switch for the new mode, the static files exist in the `media/template/( administrator || site )/templateName`. It can't be done with only the inherits value because then will not cover the parent template (or it can be done but I'm not that clever 😃)

> I am not strong in English, but for me the property name inherits is a bit of controversial

Inherits imply `inherits from ...`, that's why I picked it but naming is hard so anything else can go there, I don't have strong feelings for this

Fedik commented:
> Basically it's a switch for the new mode, the static files exist in the media/template/( administrator || site )/templateName

I see now. 
For switching we can just check if the folder exists, then do switch, something like:
```php
$template = $app->getTemplate(true);
$templatePath = JPATH_THEMES;
$client = $app->isClient('administrator') === true ? 'administrator' : 'site';

if (is_dir(JPATH_ROOT . "/media/templates/$client/{$template->template}"))
{
	$templaPath = JPATH_ROOT . "/media/templates/$client";
}
```
So if original (parent) template has the styles/scripts in /media/templates, then HTMLHelper will take it

> Inherits imply `inherits from ...`,

Exactly this part is confusing me :wink: 
I read it like `MyStyle inherits from Atum/Cassiopeia folder`, but in the implementation it like `MyStyle inherits from MyStyle folder`,

dgrammatiko commented:
> For switching we can just check if the folder exists, then do switch, something like:

This is true we could have eliminated the parent attribute both from the db and the manifest but I'm against it I mean stretching the code you could make any template `inheritable` but it will be wrong. Most, current templates use something like this:
```php
	$logo = '<img src="' . $this->baseurl . '/templates/' . $this->template . '/images/logo.svg" class="logo d-inline-block" alt="' . $sitename . '">';
```
which **will** fail miserable if you try to make a child from it (the `$this->baseurl` for the child will be `a` for the parent will be `b`, file not found will be the result)
Parent templates should always use the API for css,js,images like:
```php
$logo = HTMLHelper::image('logo.svg', '', ['class' => 'logo d-inline-block'], true, 0);
``` 
in order to not break their children...

>  read it like MyStyle inherits from Atum/Cassiopeia folder, but in the implementation it like MyStyle inherits from MyStyle folder

Are you referring to the db names or the php ones?
I might be answering the wrong questions here  😃

Fedik commented:
> Are you referring to the db names or the php ones?

To both :) 
```php
$template  =  $app->getTemplate(true);

var_dump($template->inherits); 
//   ^^^^^^^^^
// This part, I expect it like: $template inherits from "inherits value". 
// But "inherits value" contain "own folder name"

```

Well, nevermind, maybe it just me.

dgrammatiko commented:
> But "inherits value" contain "own folder name"

Maybe I misunderstood you comment in the other PR:

```
So templateDetails.xml should contain <inherits>parent_name</inherits> (or <parent_name>{blabla}</parent_name>) instead of style_id
```

Fedik commented:
> I'm against it I mean stretching the code you could make any template inheritable but it will be wrong

I think that okay. Sure, it will not work for hard-coded paths, that is not our problem.

Fedik commented:
> Maybe I misunderstood you comment in the other PR: `<inherits>parent_name</inherits>`

Yeah, I meant like this:
User create override called `foobar` that inherits `cassiopeia`, so dummy xml contain:
```xml
<name>foobar</name>
<inherits>cassiopeia</inherits>
```
Sorry for confusing :)

dgrammatiko commented:
@Fedik how about instead of `parent` we use `inheritable` (means can have children) and instead of `inherits` we use `parent` where parent is just the parent template name?

Fedik commented:
> how about instead of parent we use inheritable (means can have children) and instead of inherits we use parent where parent is just the parent template name?

That sounds good.

To be sure I right understood:

The dummy xml of child template will have something like:
```xml
<name>foobar</name>
<parent>cassiopeia</parent>
```
`<name>foobar</name>` will go to `style_template` column of database;
`<parent>cassiopeia</parent>` will go to `template` column of database;
and because we have a `parent` this will set `inheritable = 0` in database

And parent template will have `<inheritable>1</inheritable>`

I am understood correctly? :)

dgrammatiko commented:
Yup, you got it right. The pain is that I have to redo all the parent/child templates for the tests 🥱

dgrammatiko committed:
Better naming

- `parent` is  now named `inheritable`  (eg allows children)
- `inherits` is now `parent` (eg the parent template’s name )

Septdir commented:
Child themes are a great idea.
But I have one thing that confuses me.
`<inherits> siteparent </inherits>`

Usually child themes do not require their own manifest, but the name of the child theme directory indicates the link with the parent.

In the case of an example. The name of the child theme must be `siteparent_sitechild`, otherwise, to determine where is the parent and where is the child, you will have to look at the manifest of each template.

dgrammatiko commented:
> Usually child themes do not require their own manifest, but the name of the child theme directory indicates the link with the parent.

Technically a child template is an actual template but only with one file the manifest (templateDetails.xml). This way there are very few changes that need to be done in the core. Can it be done with a snake case naming? Well it's just code so it can be done but will bee extremely inefficient as you will end up scanning the templates dir and checking each folder. Now the data is coming from the db which one way or another Joomla is using it atm and has exactly 0 performance impact. Also it's not just to get something working in the core you have to think/see what's the code in com_templates. With the proposed approach the changes are minimal.

Septdir commented:
> Technically a child template is an actual template but only with one file the manifest (templateDetails.xml). This way there are very few changes that need to be done in the core. 

I recently did this for the yootheme template, I had to override a couple of classes.
The child template itself is selected in the parent's settings.

One way or another, I would still add a naming rule for the child template.
You can check during installation.

dgrammatiko commented:
> One way or another, I would still add a naming rule for the child template.

That's possible, eg prefix it with the parent name like `parentName_childName` but it's something that will be done in the com_templates and won't affect anything here

Fedik commented on installation/sql/mysql/base.sql:
```diff
@@ -793,9 +795,9 @@ CREATE TABLE IF NOT EXISTS `#__template_styles` (
-(11, 'cassiopeia', 0, '1', 'cassiopeia - Default', '{"logoFile":"","fluidContainer":"0","sidebarLeftWidth":"3","sidebarRightWidth":"3"}');
+INSERT INTO `#__template_styles` (`id`, `template`, `client_id`, `home`, `title`, `inheritable`, `parent`, `params`) VALUES
+(10, 'atum', 1, '1', 'atum - Default', '0', '0', ''),
+(11, 'cassiopeia', 0, '1', 'cassiopeia - Default', '0', '0', '{"logoFile":"","fluidContainer":"0","sidebarLeftWidth":"3","sidebarRightWidth":"3"}');
```
I think it need to be ` '0', '', ` here? For both templates

dgrammatiko replied:
Grrr, yup

dgrammatiko committed:
Fix the sql

Septdir commented:
> That's possible, eg prefix it with the parent name like `parentName_childName` but it's something that will be done in the com_templates and won't affect anything here

In the future, it will be possible to make a tree display.

But it is more necessary for developers to name their child templates correctly.

dgrammatiko commented:
@Septdir by reviewing the code is this something that breaks yootheme templates? Is this something that will be useful for you/yootheme?

Septdir commented:
> @Septdir by reviewing the code is this something that breaks yootheme templates? Is this something that will be useful for you/yootheme?

I haven't reviewed your PR code yet, and I don't think this will be a problem for YOOtheme.
In fact, you added what was missing to the core. This will be very useful, and most importantly, hello to the standardization that many developers are implementing.

My hack for YOOtheme template looks like this
https://github.com/SeptdirWorkshop/jYProExtra/blob/master/jyproextra.php#L191
And I would love to get rid of it in J4

dgrammatiko committed:
Consistency

dgrammatiko committed:
CS

Quy commented on libraries/src/Document/HtmlDocument.php:
```diff
@@ -806,13 +806,23 @@ protected function _fetchTemplate($params = array())
 		$filter = InputFilter::getInstance();
 		$template = $filter->clean($params['template'], 'cmd');
 		$file = $filter->clean($params['file'], 'cmd');
+		$inherits = $params['templateInherits'];
```
From the first page of the installer.

> Notice: Undefined index: templateInherits on line 809

dgrammatiko replied:
@Quy hmm this should have been fixed with https://github.com/joomla/joomla-cms/pull/30192/commits/7d8e760ae61efa22d748c2c403cf23d5a8db5764

Did you use the link in the description or the link at the bottom of the page to d/l?

Quy replied:
The link at the bottom of the page. I will try again once the download link is available. Thanks.

Quy replied:
`templateInherits` vs. `themeInherits` https://github.com/joomla/joomla-cms/commit/7d8e760ae61efa22d748c2c403cf23d5a8db5764

dgrammatiko replied:
https://github.com/joomla/joomla-cms/pull/30192/commits/3573ec41223f1ca5bf6979f0a14cb5c77ab3ebea solves it

Quy commented on installation/sql/postgresql/base.sql:
```diff
@@ -798,6 +798,8 @@ CREATE TABLE IF NOT EXISTS "#__template_styles" (
   "client_id" smallint DEFAULT 0 NOT NULL,
   "home" varchar(7) DEFAULT '0' NOT NULL,
   "title" varchar(255) DEFAULT '' NOT NULL,
+  `inheritable` tinyint(1) DEFAULT '0' NOT NULL,
```
```suggestion
  `inheritable`  smallint(1) DEFAULT 0 NOT NULL,
```

Quy replied:
```suggestion
  `inheritable` smallint DEFAULT 0 NOT NULL,
```

Quy commented on installation/sql/postgresql/base.sql:
```diff
@@ -808,10 +810,9 @@ CREATE INDEX "#__template_styles_idx_client_id_home" ON "#__template_styles" ("c
-(10, 'atum', 1, '1', 'atum - Default', ''),
-(11, 'cassiopeia', 0, '1', 'cassiopeia - Default', '{"logoFile":"","fluidContainer":"0","sidebarLeftWidth":"3","sidebarRightWidth":"3"}');
+INSERT INTO "#__template_styles" ("id", "template", "client_id", "home", "title", "inheritable", "parent", "params") VALUES
+(10, 'atum', 1, '1', 'atum - Default', '0', '', ''),
+(11, 'cassiopeia', 0, '1', 'cassiopeia - Default', '0', '', '{"logoFile":"","fluidContainer":"0","sidebarLeftWidth":"3","sidebarRightWidth":"3"}');
```
```suggestion
(10, 'atum', 1, '1', 'atum - Default', 0, '', ''),
(11, 'cassiopeia', 0, '1', 'cassiopeia - Default', 0, '', '{"logoFile":"","fluidContainer":"0","sidebarLeftWidth":"3","sidebarRightWidth":"3"}');
```

Quy commented on installation/sql/mysql/base.sql:
```diff
@@ -782,6 +782,8 @@ CREATE TABLE IF NOT EXISTS `#__template_styles` (
   `client_id` tinyint(1) unsigned NOT NULL DEFAULT 0,
   `home` char(7) NOT NULL DEFAULT '0',
   `title` varchar(255) NOT NULL DEFAULT '',
+  `inheritable` tinyint(1) NOT NULL DEFAULT '0',
```
```suggestion
  `inheritable` tinyint(1) NOT NULL DEFAULT 0,
```

Quy commented on installation/sql/mysql/base.sql:
```diff
@@ -793,9 +795,9 @@ CREATE TABLE IF NOT EXISTS `#__template_styles` (
-(10, 'atum', 1, '1', 'atum - Default', ''),
-(11, 'cassiopeia', 0, '1', 'cassiopeia - Default', '{"logoFile":"","fluidContainer":"0","sidebarLeftWidth":"3","sidebarRightWidth":"3"}');
+INSERT INTO `#__template_styles` (`id`, `template`, `client_id`, `home`, `title`, `inheritable`, `parent`, `params`) VALUES
+(10, 'atum', 1, '1', 'atum - Default', '0', '', ''),
+(11, 'cassiopeia', 0, '1', 'cassiopeia - Default', '0', '', '{"logoFile":"","fluidContainer":"0","sidebarLeftWidth":"3","sidebarRightWidth":"3"}');
```
```suggestion
(10, 'atum', 1, '1', 'atum - Default', 0, '', ''),
(11, 'cassiopeia', 0, '1', 'cassiopeia - Default', 0, '', '{"logoFile":"","fluidContainer":"0","sidebarLeftWidth":"3","sidebarRightWidth":"3"}');
```

Quy commented on administrator/components/com_admin/sql/updates/mysql/4.0.0-2020-06-25.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE `#__template_styles` ADD COLUMN `inheritable` tinyint(1) NOT NULL DEFAULT '0';
```
```suggestion
ALTER TABLE `#__template_styles` ADD COLUMN `inheritable` tinyint(1) NOT NULL DEFAULT 0;
```

dgrammatiko committed:
Update installation/sql/postgresql/base.sql

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Update installation/sql/postgresql/base.sql

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Update installation/sql/mysql/base.sql

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Update administrator/components/com_admin/sql/updates/mysql/4.0.0-2020-06-25.sql

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Update installation/sql/mysql/base.sql

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Update InstallationApplication.php

dgrammatiko committed:
Update InstallationApplication.php

dgrammatiko committed:
Update CMSApplication.php

dgrammatiko committed:
oops

dgrammatiko commented:
@wilsonge any signal will be good here

HLeithner commented:
We are in feature freeze for j4.0

dgrammatiko commented:
@HLeithner so then I will close this one?

HLeithner commented:
@dgrammatiko that was not my intention, it's more that we should focus on bugfixing and target new feature on 4.1 branch.

I would like to give a proper feedback on this feature too, but didn't had the time to look at the pr .

At a first glance I'm a bit confused about some decisions you made.

1. why do we need the database entry? or better why do we need both?
2. wouldn't it better to introduce a completely new "siteowner override concept", I didn't thought much about it but it's not only the template you may want to extend other parts of the cms like plugins or libraries (don't know if useful).
3. Did I saw right that you only can not have a chain of templates? base template -> company template -> site template. for example piwigo has exactly this type of template engine.
4. I'm a bit concerned about performance, every database query and every file exists check costs server resources

dgrammatiko commented:
> 2. wouldn't it better to introduce a completely new "siteowner override concept"

I don't know what you mean by that but the child templates actually solve a very real problem: updates! (but not only that)

> 1,3,4

You have to check what I did here and also step through a bit in the lifecycle. The child template is a **real** template thus the only extra load is some lookups for layouts and static assets, there is **no extra db query**. The idea of chain able templates will not work well with Joomla simply because the lookups will grow exponentially per subfolder and thus it will be awfully slow. But you can deal with it in the com_templates (eg presentation).

Fedik commented:
> I'm a bit concerned about performance, every database query and every file exists check costs server resources

It is reuse existing template style, so no extra cost

HLeithner commented:
thanks for the explanation I would like to make some code suggestion, I hope it's ok for you I have reopened the pr.
beside my comments the pr should be rebased on 4.1

HLeithner commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -394,7 +399,17 @@ public function getTemplate($params = false)
 		if (\is_object($this->template))
 		{
-			if (!file_exists(JPATH_THEMES . '/' . $this->template->template . '/index.php'))
+			if ($this->template->parent)
+			{
+				if (!file_exists(JPATH_THEMES . '/' . $this->template->template . '/index.php'))
+				{
+					if (!file_exists(JPATH_THEMES . '/' . $this->template->parent . '/index.php'))
+					{
+						throw new \InvalidArgumentException(Text::sprintf('JERROR_COULD_NOT_FIND_TEMPLATE', $this->template->template));
+					}
+				}
+			}
+			elseif (!file_exists(JPATH_THEMES . '/' . $this->template->template . '/index.php'))
```
```suggestion
			if (
				!file_exists(JPATH_THEMES . '/' . $this->template->template . '/index.php')
				&& (
					empty($this->template->parent)
					|| !file_exists(JPATH_THEMES . '/' . $this->template->parent . '/index.php')
				)
			)
```

dgrammatiko replied:
@HLeithner the rest of the changes are fine but not this one. You can reduce the inner 2 conditionals to 1 but changing it to 1 (top) conditional is wrong in the sense that the `($this->template->parent)` is extremely cheap to resolve but 
```php
!file_exists(JPATH_THEMES . '/' . $this->template->template . '/index.php')
				&& (
					empty($this->template->parent)
					|| !file_exists(JPATH_THEMES . '/' . $this->template->parent . '/index.php')
				)
```
introduces 2 more file lookups for every execution for no reason.

HLeithner replied:
> introduces 2 more file lookups for every execution for no reason.

actually not true, the second `!file_exists` will not be called if `empty($this->template->parent)` is true

but that was not the point you can leave it that way if you want it's only to reduce duplicated code. Will you merge the other suggestions?

dgrammatiko replied:
Done, although I have to state that I'm not so sure I want to move this one forward anymore.

HLeithner replied:
What's the reason? because I said it can not go into 4.1?

dgrammatiko replied:
Nothing personal don't get me wrong. It's just an idea that should have been proposed way before (my bad). The 4.1 has a lot of drawbacks for this **particular** implementation. Also not transferring the assets to the `media` folder at the time a new mode is introduced means that the templates will have to stay with the archaic way (there won't be another chance for this change in the near future). In short good intentions here but too late...

HLeithner replied:
Did I missed something? why do we have any file to the media folder and where is the b/c or drawbacks for the particular implementation?

dgrammatiko replied:
Read the documentation part in the description

HLeithner replied:
I read them but the explanation but they don't make sense to me, a js that's template independent is something different then a image that's only used in the template. We already split everything about the joomla filesystem at least the templates are "self contained" but that's only my opinion. 
Anyway that doesn't hold this feature back.

SharkyKZ replied:
The move to `media` folder seems unnecessary. Can't we just check for assets in child template directory and fallback to parent template directory?

If you want to move assets to `media`  directory for other purposes, that could be done in a separate PR so it doesn't have to wait for this.

dgrammatiko replied:
You really don’t get it. It’s you only opportunity to fix this and bring consistency...

Bakual replied:
> You really don’t get it. It’s you only opportunity to fix this and bring consistency...

Do a PR where you only move the assets to media. I actually think that could still be done for 4.0 as it's not a new feature.
If it can be done independant, lets do it independant and don't mix things.

dgrammatiko replied:
Guys do you even realise the amount of work I put here to create all the 4 templates and the other html overrides for the tests?

Asking to break it in 2 is just asking me to repeat the whole work twice. Why? This is a new mode, it doesn't  and shouldn't affect **any of the current templates**

HLeithner commented on libraries/src/Application/SiteApplication.php:
```diff
@@ -529,7 +545,35 @@ public function getTemplate($params = false)
 
 		// Fallback template
-		if (!file_exists(JPATH_THEMES . '/' . $template->template . '/index.php'))
+		if (!empty($template->parent))
```
I think you can simplify this too like them above to reduce code duplication

HLeithner commented on libraries/src/Installer/Adapter/TemplateAdapter.php:
```diff
@@ -332,13 +334,17 @@ protected function parseQueries()
 					'0',
 					Text::sprintf('JLIB_INSTALLER_DEFAULT_STYLE', Text::_($this->extension->name)),
 					$this->extension->params,
+					intval($this->manifest->inheritable),
```
```suggestion
					(int) $this->manifest->inheritable,
```
is there a reason why you not just cast to int?

HLeithner commented on libraries/src/Installer/Adapter/TemplateAdapter.php:
```diff
@@ -332,13 +334,17 @@ protected function parseQueries()
 					Text::sprintf('JLIB_INSTALLER_DEFAULT_STYLE', Text::_($this->extension->name)),
 					$this->extension->params,
+					intval($this->manifest->inheritable),
+					$this->manifest->parent ? $this->manifest->parent : '',
```
```suggestion
					$this->manifest->parent ?: '',
```

HLeithner commented on libraries/src/MVC/View/HtmlView.php:
```diff
@@ -499,8 +503,20 @@ protected function _setPath($type, $path)
-					$this->_addPath('template', $fallback);
+					$name = $this->getName();
+
+					if (!empty($template->parent))
+					{
+						// Parent template's overrides
+						$this->_addPath('template', JPATH_THEMES . "/$template->parent/html/$component/$name");
+
+						// Child template's overrides
+						$this->_addPath('template', JPATH_THEMES . "/$template->template/html/$component/$name");
+
+						break;
+					}
+
+					$this->_addPath('template', JPATH_THEMES . "/$template->template/html/$component/$name");
```
```suggestion
					if (!empty($template->parent))
					{
						// Parent template's overrides
						$this->_addPath('template', JPATH_THEMES . '/' . $template->parent . '/html/' . $component . '/' . $name);

						// Child template's overrides
						$this->_addPath('template', JPATH_THEMES . '/' . $template->template . '/html/' . $component . '/' . $name);

						break;
					}

					$this->_addPath('template', JPATH_THEMES . '/' . $template->template . '/html/' . $component . '/' . $name);
```

HLeithner reviewed (changes requested):
The PR looks ok to me.

dgrammatiko committed:
Update libraries/src/Installer/Adapter/TemplateAdapter.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

dgrammatiko committed:
Update libraries/src/Installer/Adapter/TemplateAdapter.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

dgrammatiko committed:
Update libraries/src/MVC/View/HtmlView.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

bonzani commented:
I have tested this item :white_check_mark: successfully on eec9d9adb19e81c79434bf4beeaf19fa6905dadc<br /><br />One of the most useful things in J4 I've seen so far...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30192">issues.joomla.org/tracker/joomla-cms/30192</a>.</sub>

wilsonge commented:
I really really shouldn't be pressing merge here - but I've seen several template clubs in favour of this so I'm going to break my own rules and hope I don't regret it. thankyou for this really nice feature @dgrammatiko

richard67 commented on installation/sql/postgresql/base.sql:
```diff
@@ -798,6 +798,8 @@ CREATE TABLE IF NOT EXISTS "#__template_styles" (
   "client_id" smallint DEFAULT 0 NOT NULL,
   "home" varchar(7) DEFAULT '0' NOT NULL,
   "title" varchar(255) DEFAULT '' NOT NULL,
+  `inheritable` smallint DEFAULT 0 NOT NULL,
+  `parent` varchar(50) DEFAULT '',
```
Strange that nobody saw this and nobody tested with PostgreSQL like it should always be done when a PR has database changes. This has broken installation on PostgreSQL. Will prepare a fix.

richard67 replied:
Wong names quotes I mean.

richard67 replied:
Fixed with #30327 .

richard67 commented on administrator/components/com_admin/sql/updates/postgresql/4.0.0-2020-06-25.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE "#__template_styles" ADD COLUMN "inheritable" smallint NOT NULL DEFAULT 0;
+ALTER TABLE "#__template_styles" ADD COLUMN "parent" varchar(50) DEFAULT "";
```
Next mistake `DEFAULT ""` should be  `DEFAULT ''`

alikon replied:
please submit a fix i'll test it

richard67 replied:
Done #30330 .
